jobQueue/setup.js

```diff
@@ -18,19 +18,47 @@ const conf = require('../config');
 const featureToggles = require('feature-toggles');
 const urlParser = require('url');
 const kue = require('kue');
+const activityLogUtil = require('../utils/activityLog');
+
 const redisOptions = {
   redis: conf.redis.instanceUrl.queue,
 };
+
+const express = require('express');
+const server = require('http').Server(express());
+const SHUTDOWN_TIMEOUT = 5000;
+const SUCCESSFUL_EXIT = 0;
+
 const redisInfo = urlParser.parse(redisOptions.redis, true);
 if (redisInfo.protocol !== PROTOCOL_PREFIX) {
   redisOptions.redis = 'redis:' + redisOptions.redis;
 }
 
 const jobQueue = kue.createQueue(redisOptions);
 
+/**
+ * Finishes (process.exit) when all workers tell Kue they are stopped.
+ */
+function gracefulShutdown() {
+  const start = Date.now();
+  jobQueue.shutdown(SHUTDOWN_TIMEOUT, (err) => {
+    if (featureToggles.isFeatureEnabled('enableSigtermActivityLog')) {
+      const status = 'Job queue shutdown: ' + (err || 'OK');
+      const logWrapper = {
+        status,
+        totalTime: `${Date.now() - start}ms`,
+      };
+      activityLogUtil.printActivityLogString(logWrapper, 'sigterm');
+    }
+
+    process.exit(SUCCESSFUL_EXIT);
```

Comment:
still process.exiting here?

Reply:
but job queue shutdown may not be the *only* shutdown activity -- there may be other in-flight requests which still need time to finish themselves up before exiting, right?

Reply:
Agree, as per story description I thought the solution would be now just for jobQueue, which is the only one we have now...

Do you want a generic solution? Checking open sessions, db connections?

Reply:
You can think of it as generic vs non-generic - the point is we can't be terminating the process when *other* work is still in progress

Reply:
So, the shutdown is not just related to the Kue, right?

If yes, we have to clean up resources as much as possible (db, session, etc.). At this point, if we just don't execute process.exit the server will be idle after a little while but will shutdown automatically as SIGTERM can be ignored.

Usually, there is a timeout before forcefull exiting... express js has a recommendation around that: https://expressjs.com/en/advanced/healthcheck-graceful-shutdown.html was that your first idea?

Reply:
go back to the excerpt from heroku docs which i quoted in the user story